File: 0518-coin-change-2/cpp/Solution.cpp
class Solution {
public:
    int change(int amount, vector<int> &coins) {
        int n = coins.size();
        vector<vector<int>> dp(n, vector<int>(amount + 1));
        for (int c = 0; c <= amount / coins[0]; ++c) {
            dp[0][c * coins[0]] = 1;
        }
        for (int i = 1; i < n; ++i) {
            for (int j = 0; j <= amount; ++j) {
                int k = j / coins[i];
                for (int c = 0; c <= k; ++c) {
                    dp[i][j] += dp[i - 1][j - c * coins[i]];
                }
            }
        }
        return dp[n - 1][amount];
    }
};

File: 0518-coin-change-2/python/Solution.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        dp = [[0] * (amount + 1) for _ in range(n)]
        for c in range(amount // coins[0] + 1):
            dp[0][c*coins[0]] = 1
        for i in range(1, n):
            for j in range(amount+1):
                k = j // coins[i] + 1
                for c in range(k):
                    dp[i][j] += dp[i-1][j - c * coins[i]]
        return dp[n-1][amount]
